### PR TITLE
fuse: Better check for whether snapshots changed

### PR DESCRIPTION
**Pull request by greatroar**
What does this PR change? What problem does it solve?
-----------------------------------------------------

We previously checked whether the set of snapshots might have changed based only on their number, which fails when as many snapshots are forgotten as are added. Check for the SHA-256 of their id's instead.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

No.

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

**Comment by greatroar** [comment]
Test failure is something relating to locking.

File: internal/fuse/snapshots_dirstruct.go [hunk 1]
```diff
@@ -302,10 +303,20 @@ func (d *SnapshotsDirStructure) updateSnapshots(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
-	// sort snapshots ascending by time (default order is descending)
+
+	// Sort snapshots ascending by time (default order is descending).
+	// This needs to be done before hashing.
 	sort.Sort(sort.Reverse(snapshots))
```
Comment (MichaelEischer) [inline internal/fuse/snapshots_dirstruct.go:309]: How are two snapshots with identical timestamps ordered? afaict that order isn't stable?

File: internal/fuse/snapshots_dirstruct.go [hunk 2]
```diff
@@ -302,10 +303,20 @@ func (d *SnapshotsDirStructure) updateSnapshots(ctx context.Context) error {
...
-	if d.snCount == len(snapshots) {
+	// We update the snapshots when the hash of their id's changes.
+	h := sha256.New()
+	for _, sn := range snapshots {
+		h.Write(sn.ID()[:])
+	}
+	var hash [sha256.Size]byte
+	h.Sum(hash[:0])
+
+	if d.lastCheck != (time.Time{}) && d.hash == hash {
```
Comment (MichaelEischer) [inline internal/fuse/snapshots_dirstruct.go:319]: Just checking whether the hash matches would probably be enough. Otherwise we'd have a hash collision.

Comment (greatroar) [reply internal/fuse/snapshots_dirstruct.go:319]: Yes, we'd need to have a set of snapshots with an all-0 SHA2. Removed the timestamp check.

File: internal/fuse/snapshots_dirstruct.go (hunk 1 above)
Comment (greatroar) [reply internal/fuse/snapshots_dirstruct.go]: Good point. Fixed to sort by timestamp, then hash.

**Review by MichaelEischer — APPROVED** [review]
LGTM.

Although the more detailed check only makes a difference when using `mount --no-lock`. Otherwise, we can only add new snapshots, which will necessarily change the snapshot count.

**Pull request merged**